hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/HoodieFileGroupReaderRecordReader.java

```diff
@@ -0,0 +1,279 @@
+        commitTime = FSUtils.getDeltaCommitTimeFromLogPath(realtimeSplit.getPath());
+      } else {
+        fileID = FSUtils.getFileId(realtimeSplit.getPath().getName());
+        commitTime = FSUtils.getCommitTime(realtimeSplit.getPath().toString());
+      }
+      HoodieFileGroupId fileGroupId = new HoodieFileGroupId(FSUtils.getPartitionPath(realtimeSplit.getBasePath(),
+          realtimeSplit.getPath().getParent().toString()).toString(), fileID);
+      if (isLogFile) {
+        return new FileSlice(fileGroupId, commitTime, null, realtimeSplit.getDeltaLogFiles());
+      }
+      hosts.put(realtimeSplit.getPath().toString(), realtimeSplit.getLocations());
+      HoodieBaseFile hoodieBaseFile = new HoodieBaseFile(fs.getFileStatus(realtimeSplit.getPath()), bootstrapBaseFile);
+      return new FileSlice(fileGroupId, commitTime, hoodieBaseFile, realtimeSplit.getDeltaLogFiles());
+    }
+    //cow
+    HoodieFileGroupId fileGroupId = new HoodieFileGroupId(FSUtils.getFileId(split.getPath().getName()),
+        FSUtils.getPartitionPath(tableBasePath, split.getPath().getParent().toString()).toString());
+    hosts.put(split.getPath().toString(), split.getLocations());
+    return new FileSlice(fileGroupId, FSUtils.getCommitTime(split.getPath().toString()), new HoodieBaseFile(fs.getFileStatus(split.getPath()), bootstrapBaseFile), Collections.emptyList());
+  }
+
+  private static BaseFile createBootstrapBaseFile(FileSplit split, Map<String, String[]> hosts, FileSystem fs) throws IOException {
+    if (split instanceof BootstrapBaseFileSplit) {
+      BootstrapBaseFileSplit bootstrapBaseFileSplit = (BootstrapBaseFileSplit) split;
+      FileSplit bootstrapFileSplit = bootstrapBaseFileSplit.getBootstrapFileSplit();
+      hosts.put(bootstrapFileSplit.getPath().toString(), bootstrapFileSplit.getLocations());
+      return new BaseFile(fs.getFileStatus(bootstrapFileSplit.getPath()));
+    }
+    return null;
+  }
+
+  private static Schema createRequestedSchema(Schema tableSchema, JobConf jobConf) {
+    String readCols = jobConf.get(ColumnProjectionUtils.READ_COLUMN_NAMES_CONF_STR);
+    if (StringUtils.isNullOrEmpty(readCols)) {
+      Schema emptySchema = Schema.createRecord(tableSchema.getName(), tableSchema.getDoc(),
+          tableSchema.getNamespace(), tableSchema.isError());
+      emptySchema.setFields(Collections.emptyList());
+      return emptySchema;
+    }
+    //hive will handle the partition cols
+    String partitionColString = jobConf.get(hive_metastoreConstants.META_TABLE_PARTITION_COLUMNS);
+    Set<String> partitionColumns;
+    if (partitionColString == null) {
+      partitionColumns = Collections.emptySet();
+    } else {
+      partitionColumns = Arrays.stream(partitionColString.split(",")).collect(Collectors.toSet());
+    }
+    //if they are actually written to the file, then it is ok to read them from the file
+    tableSchema.getFields().forEach(f -> partitionColumns.remove(f.name().toLowerCase(Locale.ROOT)));
```

Comment:
I'm confused, will partitionColumns always be empty?

Reply:
partitionColumns won't be empty if any of the partition columns are not written to the file. tableSchema only has the columns that are written to the file. This is the case in the docker demo https://hudi.apache.org/docs/docker_demo#step-3-sync-with-hive . If you look at the data in https://github.com/apache/hudi/tree/master/docker/demo/data, there is no field named "dt".